cmd/clusterctl/client/cluster/ca_injection.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	return string(ca), nil
+}
+
+// splitObjectKey splits the string by the name separator and returns it as client.ObjectKey.
+func splitObjectKey(nameStr string) client.ObjectKey {
+	splitPoint := strings.IndexRune(nameStr, types.Separator)
+	if splitPoint == -1 {
+		return client.ObjectKey{Name: nameStr}
+	}
+	return client.ObjectKey{Namespace: nameStr[:splitPoint], Name: nameStr[splitPoint+1:]}
+}
+
+// checkObject gets the desired CA certificate and compares it relevant field on
+// the object which is first read from kube-apiserver.
+func (c *certificateInjectionVerifier) checkObject(ctx context.Context, obj unstructured.Unstructured) error {
+	// get the CA certificate from the Certificate's secret which is referred in an
+	// annotation at the object.
+	ca, err := c.getCACertificateFor(ctx, obj)
+	if err != nil {
+		return err
+	}
+
+	// Build the object key from the passed object.
+	objKey := client.ObjectKey{Namespace: obj.GetNamespace(), Name: obj.GetName()}
+
+	// Get the object and assert the CA certificate.
+	switch obj.GetKind() {
+	case customResourceDefinitionKind:
+		crd := &apiextensionsv1.CustomResourceDefinition{}
+		if err := c.Client.Get(ctx, objKey, crd); err != nil {
+			return err
+		}
+
+		if crd.Spec.Conversion.Webhook == nil || crd.Spec.Conversion.Webhook.ClientConfig == nil || string(crd.Spec.Conversion.Webhook.ClientConfig.CABundle) != ca {
+			return fmt.Errorf("injected CA for CustomResourceDefinition %s does not match", objKey)
+		}
+	case mutatingWebhookConfigurationKind:
+		mutatingWebhook := &admissionregistrationv1.MutatingWebhookConfiguration{}
+		if err := c.Client.Get(ctx, objKey, mutatingWebhook); err != nil {
+			return err
+		}
+		for _, webhook := range mutatingWebhook.Webhooks {
+			if string(webhook.ClientConfig.CABundle) != ca {
+				return fmt.Errorf("injected CA for MutatingWebhookConfiguration %s does not match", objKey)
+			}
+		}
+	case validatingWebhookConfigurationKind:
+		validatingWebhook := &admissionregistrationv1.ValidatingWebhookConfiguration{}
+		if err := c.Client.Get(ctx, objKey, validatingWebhook); err != nil {
+			return err
+		}
+		for _, webhook := range validatingWebhook.Webhooks {
+			if string(webhook.ClientConfig.CABundle) != ca {
+				return fmt.Errorf("injected CA for ValidatingWebhookConfiguration %s does not match", objKey)
+			}
+		}
+	default:
+		return fmt.Errorf("unknown object type %s", obj.GetKind())
+	}
+
+	return nil
+}
+
+// newVerifyBackoff creates a new API Machinery backoff parameter set suitable for use with clusterctl verify operations.
+func newVerifyBackoff() wait.Backoff {
+	// Return a exponential backoff configuration which returns durations for a total time of ~5m.
+	// Example: 0, .5s, 1.2s, 2.3s, 4s, 6s, 10s, 16s, 24s, 37s, 57s, 85s, 129s, 194s, 291s
+	// Jitter is added as a random fraction of the duration multiplied by the jitter factor.
+	return wait.Backoff{
```

Comment:
Should we make this shorter?

5m seems like a very long time

Reply:
Modified to be part of wait-for-providers and for each provider it is included now in the `wait-provider-timeout`.